<a href="https://colab.research.google.com/github/KwakuBonfulBosompim/MSc-Data-Analytics-and-ML-Projects/blob/main/Bill_Authentication_using_XGBoost.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This project is about detecting fake or real banknotes using a smart machine learning algorithm called XGBoost

XGBoost (eXtreme Gradient Boosting) works like a team of tiny decision-making robots (trees). Each robot tries to fix the mistakes of the previous one, making the final prediction fast, accurate, and reliable.

this model is
Very fast on big data,
Handles missing values automatically,
Can prevent overfitting (like memorizing the training data instead of learning patterns),
Very accurate, often winning competitions


Now we going to use a dataset named bill_authentication.csv for the implementation of the XGBoost model


CSV file with features: size, number of rooms, location, age
And the target: price

Training set → teach the model
Test set → check if the model learned correctly


Tune Parameters ⚙️
You can adjust “how many trees,” “how deep each tree is,” or “learning rate” to make it smarter

First we import the libraries and load the dataset

In [1]:
import pandas as pd
import numpy as np

In [3]:
data = pd.read_csv('bill_authentication.csv')

# techniques to check and access the datasets
data.head()

Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

In [4]:
data.info

<bound method DataFrame.info of       Variance  Skewness  Curtosis  Entropy  Class
0      3.62160   8.66610   -2.8073 -0.44699      0
1      4.54590   8.16740   -2.4586 -1.46210      0
2      3.86600  -2.63830    1.9242  0.10645      0
3      3.45660   9.52280   -4.0112 -3.59440      0
4      0.32924  -4.45520    4.5718 -0.98880      0
...        ...       ...       ...      ...    ...
1367   0.40614   1.34920   -1.4501 -0.55949      1
1368  -1.38870  -4.87730    6.4774  0.34179      1
1369  -3.75030 -13.45860   17.5932 -2.77710      1
1370  -3.56370  -8.38270   12.3930 -1.28230      1
1371  -2.54190  -0.65804    2.6842  1.19520      1

[1372 rows x 5 columns]>

In [7]:
X = data.iloc[:,:1].values
X[:5]

array([[3.6216 ],
       [4.5459 ],
       [3.866  ],
       [3.4566 ],
       [0.32924]])

In [10]:
y = data.iloc[:, -1].values
y[:17]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Now we split the data

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

we import the XGboost library

it could be used for classification or regression problem or other tasks


Our Project is a Classification Problem
whether a bank note is forged or not

In [12]:
from xgboost import XGBClassifier

Now We build the Algorithm / Specify the parameters
 n_estimators : Optional[int]
        Number of boosting rounds.
    max_depth :  typing.Optional[int]
        Maximum tree depth for base learners

In [15]:
bst = XGBClassifier(n_estimators=200, max_depth=3, learning_rate=0.5, objective='binary:logistic')

Now we train the model based on the data (Fit the model to the data)

In [16]:
bst.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              feature_weights=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=0.5, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=3,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=200,
              n_jobs=None, num_parallel_tree=None, ...)

Now we Estimate / make the prediction

In [17]:
y_pred = bst.predict(X_test)

In [19]:
y_pred[:5]

array([0, 0, 1, 0, 0])

Now We check accurate the model predicted the target value based on the train data we provided

In [20]:
from sklearn.metrics import confusion_matrix, classification_report

In [21]:
print(confusion_matrix(y_test, y_pred))

[[142  24]
 [ 28  81]]


[[ True Positive, False Negative ],
 [ False Positive, True Negative ]]

142 → Real bills correctly detected ✅

24 → Real bills incorrectly labeled as fake ❌

28 → Fake bills incorrectly labeled as real ❌ (this is risky!)

81 → Fake bills correctly detected

**Now we try to optimize the model
approach to ensure we create or adopt the necessary parameters**


**Optimization is like tuning your robot’s “settings or configuration"**

three common approaches

**Manual tuning 🎛️**
You change one parameter at a time and see if accuracy improves.
Example: Increase n_estimators from 100 → 200 → 300 and check results.
Simple, but slow

**Grid Search 🔍**
You define a grid of parameter values:
The computer tries all combinations → picks the best one.
Accurate but can take a long time for many parameters


**Randomized Search 🎲**
Like Grid Search, but instead of trying all combinations, it samples randomly.
Faster for big datasets, still usually finds good settings

**Bayesian Optimization (advanced)**
The computer learns which settings are better over time instead of trying everything.
Very efficient, but more advanced

In [22]:
from sklearn.model_selection import GridSearchCV

In [23]:
# GridSearchCV with XGBoost

param_grid = {'n_estimators': [150, 250, 250],
              'learning_rate': [0.1, 0.5, 1]}
# Total combinations = 3 × 3 = 9 models
# if you use 5-fold cross-validation (cv=5):
# Total models = 9 × 5 = 45 models trained.

In [24]:
# Now we create our grid
mygrid = GridSearchCV(XGBClassifier(), param_grid, cv=5)

In [25]:
mygrid.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=XGBClassifier(base_score=None, booster=None,
                                     callbacks=None, colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None, device=None,
                                     early_stopping_rounds=None,
                                     enable_categorical=False, eval_metric=None,
                                     feature_types=None, feature_weights=None,
                                     gamma=None, grow_policy=None,
                                     importance_type=None,
                                     interaction_constraints=None,
                                     learning_rate=None, max_bin=None,
                                     max_cat_threshold=None,
                                     max_cat_to_onehot=None,
                                     max_delta_step=None, max_depth=None,
                                     max_leaves=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     multi_strategy=None, n_estimators=None,
                                     n_jobs=None, num_parallel_tree=None, ...),
             param_grid={'learning_rate': [0.1, 0.5, 1],
                         'n_estimators': [150, 250, 250]})

In [26]:
mygrid.best_params_

{'learning_rate': 0.1, 'n_estimators': 150}

In [27]:
mygrid.best_score_

np.float64(0.8359402241594023)

Now we create the full summary of full validation

all results summarize in this table

In [28]:
mygrid_summary = pd.DataFrame(mygrid.cv_results_)

In [29]:
mygrid_summary.info

<bound method DataFrame.info of    mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.047478      0.001229         0.002989        0.000059   
1       0.071722      0.005397         0.003385        0.000099   
2       0.078162      0.008751         0.003281        0.000044   
3       0.043420      0.002776         0.003123        0.000589   
4       0.066215      0.001359         0.003258        0.000033   
5       0.072411      0.010113         0.003493        0.000338   
6       0.043932      0.001822         0.002986        0.000316   
7       0.066665      0.000963         0.003292        0.000030   
8       0.072605      0.011347         0.003328        0.000066   

   param_learning_rate  param_n_estimators  \
0                  0.1                 150   
1                  0.1                 250   
2                  0.1                 250   
3                  0.5                 150   
4                  0.5                 250   
5                  0.5                 250   
6                  1.0                 150   
7                  1.0                 250   
8                  1.0                 250   

                                        params  split0_test_score  \
0  {'learning_rate': 0.1, 'n_estimators': 150}           0.831818   
1  {'learning_rate': 0.1, 'n_estimators': 250}           0.827273   
2  {'learning_rate': 0.1, 'n_estimators': 250}           0.827273   
3  {'learning_rate': 0.5, 'n_estimators': 150}           0.822727   
4  {'learning_rate': 0.5, 'n_estimators': 250}           0.822727   
5  {'learning_rate': 0.5, 'n_estimators': 250}           0.822727   
6    {'learning_rate': 1, 'n_estimators': 150}           0.831818   
7    {'learning_rate': 1, 'n_estimators': 250}           0.831818   
8    {'learning_rate': 1, 'n_estimators': 250}           0.831818   

   split1_test_score  split2_test_score  split3_test_score  split4_test_score  \
0           0.813636           0.808219           0.858447           0.867580   
1           0.813636           0.808219           0.867580           0.858447   
2           0.813636           0.808219           0.867580           0.858447   
3           0.809091           0.808219           0.853881           0.863014   
4           0.809091           0.812785           0.844749           0.863014   
5           0.809091           0.812785           0.844749           0.863014   
6           0.818182           0.803653           0.853881           0.863014   
7           0.813636           0.808219           0.853881           0.863014   
8           0.813636           0.808219           0.853881           0.863014   

   mean_test_score  std_test_score  rank_test_score  
0         0.835940        0.023624                1  
1         0.835031        0.023852                2  
2         0.835031        0.023852                2  
3         0.831386        0.022870                7  
4         0.830473        0.020458                8  
5         0.830473        0.020458                8  
6         0.834110        0.021968                6  
7         0.834114        0.021546                4  
8         0.834114        0.021546                4  >

In [30]:
mygrid_summary.iloc[:,4]

0    0.1
1    0.1
2    0.1
3    0.5
4    0.5
5    0.5
6    1.0
7    1.0
8    1.0
Name: param_learning_rate, dtype: float64

The best model

In [31]:
mygrid_summary.iloc[:,14]

0    1
1    2
2    2
3    7
4    8
5    8
6    6
7    4
8    4
Name: rank_test_score, dtype: int32

Scores now

In [32]:
mean_test_score = mygrid_summary.iloc[:,12]

The Faults

In [33]:
# Cross Validation Fault
mygrid_summary.iloc[:, 7:12]

split0_test_score  split1_test_score  split2_test_score  split3_test_score  \
0           0.831818           0.813636           0.808219           0.858447   
1           0.827273           0.813636           0.808219           0.867580   
2           0.827273           0.813636           0.808219           0.867580   
3           0.822727           0.809091           0.808219           0.853881   
4           0.822727           0.809091           0.812785           0.844749   
5           0.822727           0.809091           0.812785           0.844749   
6           0.831818           0.818182           0.803653           0.853881   
7           0.831818           0.813636           0.808219           0.853881   
8           0.831818           0.813636           0.808219           0.853881   

   split4_test_score  
0           0.867580  
1           0.858447  
2           0.858447  
3           0.863014  
4           0.863014  
5           0.863014  
6           0.863014  
7           0.863014  
8           0.863014

Impact of the Project 🌟

Helps banks and businesses detect fake bills quickly.

Reduces human errors in bill authentication.

Shows how machine learning can automate important security tasks

Insights Achieved 🔍

XGBoost is fast, accurate, and handles missing data automatically.

Tuning parameters significantly improves model performance.

Confusion matrix helps understand mistakes and improve the model.

Even a simple dataset can be used to train a powerful model for real-world problems.

What We Achieved 🏁

Built an XGBoost model to detect real vs fake bills.

Optimized parameters using Grid Search for best accuracy.

Successfully demonstrated machine learning in bill authentication.

Created a clear workflow from dataset → model → evaluation → optimization.